zadanie 1
a) Podziel w losowy sposób bazę danych na zbiór treningowy (70%) i testowy
(30%).

In [60]:
import pandas as pd
from sklearn.model_selection import train_test_split

df = pd.read_csv("iris.csv")

X = df.iloc[:, :-1].values
y = df.iloc[:, 4].values

#podzial na zbior testowy (30%) i treningowy (70%), ziarno losowosci = 274929
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.7, random_state=274929)

b) Uruchom każdy z klasyfikatorów wykorzystując paczki i dokonaj ewaluacji ma
zbiorze testowym wyświetlając procentową dokładność i macierz błędu.

1. k-NN, k=3

In [70]:
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import classification_report, confusion_matrix

scaler = StandardScaler()
scaler.fit(X_train)

X_train = scaler.transform(X_train)
X_test = scaler.transform(X_test)

classifier = KNeighborsClassifier(n_neighbors=3)
classifier.fit(X_train, y_train)

y_pred = classifier.predict(X_test)

print(confusion_matrix(y_test, y_pred))

numberOfAll = X_test.shape[0]
numberOfWrong = (y_test != y_pred).sum()
numberOfCorrect = numberOfAll - numberOfWrong

print("Procent dobrze sklasyfikowanych irysów %d" % (numberOfCorrect/numberOfAll * 100))

[[16  1  0]
 [ 0 14  1]
 [ 0  1 12]]
Procent dobrze sklasyfikowanych irysów 93


Otrzymany wynik to 93%

2. k-NN, k=5

In [71]:
classifier = KNeighborsClassifier(n_neighbors=5)
classifier.fit(X_train, y_train)

y_pred = classifier.predict(X_test)

print(confusion_matrix(y_test, y_pred))

numberOfAll = X_test.shape[0]
numberOfWrong = (y_test != y_pred).sum()
numberOfCorrect = numberOfAll - numberOfWrong

print("Procent dobrze sklasyfikowanych irysów %d" % (numberOfCorrect/numberOfAll * 100))

[[17  0  0]
 [ 0 14  1]
 [ 0  1 12]]
Procent dobrze sklasyfikowanych irysów 95


Otrzymany wynik to 95%

3. k-NN, k=11

In [72]:
classifier = KNeighborsClassifier(n_neighbors=11)
classifier.fit(X_train, y_train)

y_pred = classifier.predict(X_test)

print(confusion_matrix(y_test, y_pred))

numberOfAll = X_test.shape[0]
numberOfWrong = (y_test != y_pred).sum()
numberOfCorrect = numberOfAll - numberOfWrong

print("Procent dobrze sklasyfikowanych irysów %d" % (numberOfCorrect/numberOfAll * 100))

[[17  0  0]
 [ 0 12  3]
 [ 0  2 11]]
Procent dobrze sklasyfikowanych irysów 88


Otrzymany wynik to 88%

zadanie 2
Dokonaj klasyfikacji zbioru danych z irysami za pomocą Naive Bayesa

In [3]:
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB

df = pd.read_csv("iris.csv")

X = df.iloc[:, :-1].values
y = df.iloc[:, 4].values

X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.7, random_state=274929)

gnb = GaussianNB()
y_pred = gnb.fit(X_train, y_train).predict(X_test)

numberOfAll = X_test.shape[0]
numberOfWrong = (y_test != y_pred).sum()
numberOfCorrect = numberOfAll - numberOfWrong

print("Procent %d"% (numberOfCorrect/numberOfAll * 100))

Procent 91


Wynik: 91% irysów zostało dobrze sklasyfikowanych.

zad 3
a) Załadować paczkę sklearn , bazę danych z irysami I podzielić irysy na część
testową i treningową używając komendy: train_test_split ( 70% / 30% ).

In [21]:
df = pd.read_csv("iris.csv")

X = df.iloc[:, :-1].values
y = df.iloc[:, 4].values

X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.7, random_state=274999)

b) Sieć neuronowa nie akceptuje napisów, jedynie liczby. Może się nauczyć
rozpoznawać irysy, pod warunkiem że ich nazwy są numeryczne np. 0, 1 lub 2.
Trzeba więc zamienić nazwy tekstowe na liczby przed treningiem.

In [22]:
target = df[['class']].replace(['setosa','versicolor','virginica'],[0,1,2])
# print(target.sample(n=5))

Jakie to liczby? Jakim napisom odpowiadają?
* setosa - 0
* versicolor - 1
* virginica - 2

c) Możesz przeskalować dane, ale ten punkt nie jest obowiązkowy

In [23]:
df_norm = df[['sepallength', 'sepalwidth', 'petallength', 'petalwidth']].apply(lambda x: (x - x.min()) / (x.max() - x.min()))

d) Skonstruuj I wytrenuj model sieci neuronowej z czteroneuronową warstwą
wejściową (bo są cztery pomiary irysów), jedną ukrytą warstwą z dwoma
neuronami i warstwą wyjściową z jednym neuronem decydującym o gatunku
irysa.

In [24]:
import pandas as pd
from sklearn.neural_network import MLPClassifier # neural network
from sklearn import metrics

df = pd.concat([df_norm, target], axis=1)
# df.sample(n=5)

train, test = train_test_split(df, train_size=0.7, random_state=274929)

trainX = train[['sepallength','sepalwidth','petallength','petalwidth']]# taking the training data features

trainY = train['class'] # output of our training data

testX= test[['sepallength','sepalwidth','petallength','petalwidth']] # taking test data features

testY = test['class'] # output value of test data

clf = MLPClassifier(solver='lbfgs', alpha=1e-5, hidden_layer_sizes=(2,), random_state=1)

clf.fit(trainX, trainY)

MLPClassifier(alpha=1e-05, hidden_layer_sizes=(2,), random_state=1,
              solver='lbfgs')

e) Wyjaśnij jak liczby zwracane przez sieć (są to liczby rzeczywiste,
zmiennoprzecinkowe) zamieniane są na odpowiedzi (gatunki irysa).

wynik jest zaokrąglany.


In [35]:
prediction = clf.predict(testX)
# print(testX)
# print(testY)
print("-")
print(prediction)

-
[2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2]


f) Dokonaj ewaluacji sieci na zbiorze testowym. Wyświetl jej dokładność.

In [36]:
print('The accuracy of the Multi-layer Perceptron is:',metrics.accuracy_score(prediction,testY))

The accuracy of the Multi-layer Perceptron is: 0.28888888888888886


Wyszło 28.8%

g) Sprawdź czy model sieci z trzema neuronami działa lepiej.

In [38]:
clf = MLPClassifier(solver='lbfgs', alpha=1e-5, hidden_layer_sizes=(3,), max_iter=1000, random_state=1)
clf.fit(trainX, trainY)



prediction = clf.predict(testX)
# print(prediction)

print('The accuracy of the Multi-layer Perceptron is:',metrics.accuracy_score(prediction,testY))

The accuracy of the Multi-layer Perceptron is: 0.8888888888888888


Działa lepiej, wyszło 88.9%

h) Sprawdź czy model z dwiema warstwami neuronowymi, po 3 neurony każda,
działa lepiej.

In [39]:
clf = MLPClassifier(solver='lbfgs', alpha=1e-5, hidden_layer_sizes=(3, 3), random_state=1)

clf.fit(trainX, trainY)

prediction = clf.predict(testX)
print(prediction)

print('The accuracy of the Multi-layer Perceptron is:',metrics.accuracy_score(prediction,testY))

[1 1 0 1 0 1 1 2 2 0 0 1 0 0 1 1 2 2 2 2 2 2 1 1 2 0 1 1 0 2 1 0 0 2 1 0 0
 0 0 0 1 0 0 2 1]
The accuracy of the Multi-layer Perceptron is: 0.9333333333333333


Działa jeszcze lepiej, wyszło: 93.4%

i) Innym sposobem odgadywania irysów jest przyporządkowanie każdej wartości
klasy jednemu neuronowi wyjściowemu (np. górny to setoa, środkowy
versicolor, dolny virginica).

In [41]:
clf = MLPClassifier(solver='lbfgs', alpha=1e-5, hidden_layer_sizes=(3,), max_iter=1000, random_state=1)

clf.fit(trainX, trainY)

prediction = clf.predict(testX)
# print(prediction)

print('The accuracy of the Multi-layer Perceptron is:',metrics.accuracy_score(prediction,testY))

The accuracy of the Multi-layer Perceptron is: 0.8888888888888888
